### PR TITLE
Add tool-chain script to automate release

### PR DESCRIPTION
**Pull request by MOmarMiraj**
This MR will create a two step release process that you can run on any branch and release the Go SDK.

This requires GitLab CLI installed and the MakeFile checks if you have it or not.

Also, a PAT will have to be created and exported to GITHUB_CLI_TOKEN environment variable to run the gitlab CLI. The  minimum required scopes for the token are: repo, read:org, and gist.

The first script `prep-release.sh` will update the version/build numbers and create a new changelog for the release

The second script `release.sh` will create the release branch, add all necessary files, and create the release with the correct changelog and version

This process will ensure that the developer can verify the files before pushing a release.

A README is in the `internal/changelogs` folder on step by step process on how to push a release.

Build Number:
image
Build Number is updated on Logs:
image

## To Test
1. Go into the test go sdk repo (https://github.com/MOmarMiraj/onepassword-sdk-go). I sent you three all invites to access the repo which should give you the correct permissions.
3. Test on main/feature branch.. run `make prep-release`.. follow instructions
4. Run `make release`

I have updated the fork repo to include the latest changes and edited it so the release will be created on the fork repo not the actual repo.. so just run the scripts and should be good to go.

### CODE REVIEW

File: scripts/go_release.sh [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+done
+
+git tag -a -s  "v${version_number}" -m "${version_number}"
+git status
+git add internal/core.go 
+git commit -m "Release for ${version_number}"
+git push 
+
+# Login with Github CLI
+gh auth login --with-token <<< ${GIT_TOKEN}
```
Comment (hculea) [inline scripts/go_release.sh:72]: Let's validate that this is in the environment variable.

Additionally, I think it should be `GITHUB_TOKEN`, to disambiguate it from git-related credentials.

File: scripts/go_release.sh [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+# Read multiline input from the user until Ctrl+D is pressed
+while IFS= read -r line; do
+    changelog_content+="$line"$'\n' # Append each line to the variable with a newline character
+done
+
+git tag -a -s  "v${version_number}" -m "${version_number}"
+git status
+git add internal/core.go 
+git commit -m "Release for ${version_number}"
+git push
```
Comment (hculea) [inline scripts/go_release.sh:69]: This would only push commits I believe. To push tags, you'd have to do `git push --tags` or `git push origin tag <tag-name>`.

File: scripts/go_release.sh [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+git tag -a -s  "v${version_number}" -m "${version_number}"
+git status
+git add internal/core.go 
+git commit -m "Release for ${version_number}"
+git push 
+
+# Login with Github CLI
+gh auth login --with-token <<< ${GIT_TOKEN} 
+
+gh release create "${version_number}" --title "Release ${version_number}" --notes "${changelog_content}" --repo github.com/1Password/onepassword-sdk-go
```
Comment (hculea) [inline scripts/go_release.sh:74]: Do you have a personal project where this could be tested functionally?

File: scripts/go_release.sh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+done
+
+# Replace version number with new version number
+awk -v build="$build_number" -v version="$version_number" '
+  /SDKSemverVersion/ && !done {
+    print "   SDKSemverVersion      =", "\"" build "\" // v" version;
+    done = 1;
+    next;
+  } 
+  { print }' internal/core.go > tmpfile && mv tmpfile internal/core.go
```
Comment (hculea) [inline scripts/go_release.sh:54]: This should not be necessary.

We could simply use a `version` file, and `go:embed` this into the `SDKSemverVersion` variable. Let me know if you'd like to pair up on this.

File: scripts/go_release.sh (hunk 2 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh:69]: This also pushes tags, this was tested on my fork of the repo.

File: scripts/go_release.sh (hunk 3 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh:74]: Tested on the fork of the Go SDK repo where it successfully created the release.

File: scripts/go_release.sh [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+# Read multiline input from the user until Ctrl+D is pressed
+while IFS= read -r line; do
+    changelog_content+="$line"$'\n' # Append each line to the variable with a newline character
+done
+
+git tag -a -s  "v${version_number}" -m "${version_number}"
+git status
+git add internal/core.go 
+git commit -m "Release for ${version_number}"
```
[the remark this answers is not in the document]
Comment (MOmarMiraj) [reply scripts/go_release.sh]: Done!

File: scripts/go_release.sh (hunk 4 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh]: Done!

File: scripts/go_release.sh [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+git push 
+
+# Login with Github CLI
+gh auth login --with-token <<< ${GIT_TOKEN} 
+
+gh release create "${version_number}" --title "Release ${version_number}" --notes "${changelog_content}" --repo github.com/1Password/onepassword-sdk-go
+ 
+
+
```
[the remark this answers is not in the document]
Comment (MOmarMiraj) [reply scripts/go_release.sh]: Fixed!

File: internal/core.go [hunk 7]
```diff
@@ -2,15 +2,18 @@ package internal
...
 	"runtime"
 )
 
 const (
-	SDKSemverVersion      = "0010009" // v0.1.0-beta.9
 	SDKLanguage           = "Go"
 	DefaultRequestLibrary = "net/http"
 )
 
+//go:embed version-build.txt
```
Comment (hculea) [inline internal/core.go:14]: nit: I don't think the `.txt` extension is necessary, here. I don't see any immediate disadvantages to it, but naming the file simply `version-build` may be cleaner.

File: scripts/go_release.sh [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        return 0
+    else
+        echo "Invalid build number format: ${build}"
+        return 1
+    fi
+}
+
+# Validate the version number and build number from their respective files
+if ! validate_version_number "$version_number" || ! validate_build_number "$build_number"; then
+    exit 1
```
Comment (hculea) [inline scripts/go_release.sh:37]: We should ensure that there is a clear error message here, before exiting early. May even want to split the two cases (version/build number are invalid) for better debug ability.

File: scripts/go_release.sh [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+# Prompt the user to input multiline text
+echo "Enter your multiline text (press Ctrl+D when finished):"
+changelog_content=""
+
+# Read multiline input from the user until Ctrl+D is pressed
+while IFS= read -r line; do
+    changelog_content+="${line}"$'\n' # Append each line to the variable with a newline character
+done
+
+git tag -a -s  "v${version_number}" -m "${version_number}"
```
Comment (hculea) [inline scripts/go_release.sh:49]: I don't think the tag that you create here is ever pushed, is it?

`git push` by default should definitely not be pushing tags by default. 

Resource: "Pushing tags to remote" section [here](https://www.atlassian.com/git/tutorials/inspecting-a-repository/git-tag)

File: scripts/go_release.sh [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+# Add changes and commit/push to branch
+git add .
+git commit -m "Release for ${version_number}"
+git push origin ${branch}
+
+# Login with Github CLI
+gh auth login --with-token <<< ${GITHUB_TOKEN} 
+
+gh release create "${version_number}" --title "Release ${version_number}" --notes "${changelog_content}" --repo github.com/MOmarMiraj/onepassword-sdk-go
```
Comment (hculea) [inline scripts/go_release.sh:70]: restating my feedback from the previous review: we should validate that `GITHUB_TOKEN` is in the environment before proceeding to the script.

Currently, if no GITHUB_TOKEN is set, the script will do everything up to the tagging but without the release, leaving us in an inconsistent state.

File: scripts/go_release.sh [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    else
+        echo "Invalid version number format: ${version}"
+        return 1
+    fi
+}
+
+# Function to validate the build number format Mmmppbb
+validate_build_number() {
+    local build="$1"
+    if [[ "${build}" =~ ^[0-9]{1}[0-9]{2}[0-9]{2}[0-9]{2}$ ]]; then
```
Comment (hculea) [inline scripts/go_release.sh:26]: can `[0-9]{1}[0-9]{2}[0-9]{2}[0-9]{2}` simply be `[0-9]{7}`? I understand the intention of making this clearer, but I think a comment explaining what the 7 digits are would be better. Regex patterns can be intimidating

File: scripts/go_release.sh
Comment (hculea) [inline scripts/go_release.sh:1]: nit: no need to call this `go_release`. This is a Go SDK, with exclusively Go code within. We can simply call this `release.sh`.

File: Makefile [hunk 12]
```diff
@@ -0,0 +1,10 @@
+
+check_if_gh_is_installed:
+	@if ! command -v gh &> /dev/null; then\
+		echo "gh is not installed";\
+		exit 1;\
+	fi
```
Comment (hculea) [inline Makefile:6]: I would prefer this moved to the script, to keep the make file clean. As we add more validation (e.g. for `GITHUB_TOKEN`, as I commented before) it's better to keep it centralized.

File: Makefile [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+
+release:
+	make check_if_gh_is_installed
+	scripts/go_release.sh
```
Comment (hculea) [inline Makefile:10]: nit: let's add a proper newline here

File: scripts/go_release.sh (hunk 8 above)
Comment (hculea) [reply scripts/go_release.sh:37]: Oh wait, I am seeing this has changed - is the expectation here that the user changes the build and version number in the files before running the script? If so, we should document this, but I wonder if we have any reason for not asking for these interactively?

File: scripts/go_release.sh [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    fi
+}
+
+# Validate the version number and build number from their respective files
+if ! validate_version_number "$version_number" || ! validate_build_number "$build_number"; then
+    exit 1
+fi
+
+# Prompt the user to input multiline text
+echo "Enter your multiline text (press Ctrl+D when finished):"
```
Comment (hculea) [inline scripts/go_release.sh:41]: Let's make sure the user knows what they are prompted for, here. "multiline text" is not very self-explanatory

File: scripts/go_release.sh [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+# Function to validate the build number format Mmmppbb
+validate_build_number() {
+    local build="$1"
+    if [[ "${build}" =~ ^[0-9]{1}[0-9]{2}[0-9]{2}[0-9]{2}$ ]]; then
+        echo "Updated build number is: ${build}"
+        return 0
+    else
+        echo "Invalid build number format: ${build}"
```
Comment (hculea) [inline scripts/go_release.sh:30]: pro-tip - you can ask the user directly to edit the version and build files in the script by running something like `${EDITOR:-nano} "$LOCATION_OF_FILE"

**Review by hculea — CHANGES_REQUESTED** [review]

File: scripts/go_release.sh (hunk 8 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh:37]: I changed it as it didn't feel good having to change the file version/build numbers and then going into the script to ask the user for the version/build numbers, could increase the chance of error and if we are updating it at one spot.. its better to just grab it from there instead of asking it twice.

File: scripts/go_release.sh (hunk 10 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh:70]: GITHUB_TOKEN is actually used by git for authentication, as I receive this error...
`The value of the GITHUB_TOKEN environment variable is being used for authentication.
To have GitHub CLI store credentials instead, first clear the value from the environment.`

I will change the variable name to GITHUB_CLI_TOKEN to be more clear and not interfere with git.

File: scripts/go_release.sh (hunk 15 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh]: Awesome, I changed it a bit to make sure the user inputs it correctly and not just error out.

File: scripts/go_release.sh (hunk 14 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh]: Done!

File: Makefile (hunk 13 above)
Comment (MOmarMiraj) [reply Makefile]: Done!

File: Makefile (hunk 12 above)
Comment (MOmarMiraj) [reply Makefile]: Moved and done!

File: scripts/go_release.sh
Comment (MOmarMiraj) [reply scripts/go_release.sh:1]: Done!

File: scripts/go_release.sh (hunk 11 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh]: Added a comment to explain the build number format and changed regex.

File: internal/core.go (hunk 7 above)
Comment (MOmarMiraj) [reply internal/core.go:14]: Done.

File: scripts/go_release.sh (hunk 10 above)
Comment (hculea) [reply scripts/go_release.sh]: I actually prefer prefer simply setting the `GITHUB_TOKEN` and avoiding the `gh auth login --with-token <<< ${GITHUB_TOKEN} ` line - it's shorter and more concise. It's a bit redundant to use another command to auth, since we're passing the cred by envvar anyway.

File: internal/core.go (hunk 7 above)
Comment (hculea) [reply internal/core.go:14]: I don't think this is done

File: scripts/go_release.sh (hunk 8 above)
Comment (hculea) [reply scripts/go_release.sh]: You're correct - could make a lot of sense to create a script that updates these too, as well as fetches the "changelog" (release notes). e.g. `prepare-release` vs actual `release` (which does the tagging)

But I'll defer this decision to you

File: scripts/go_release.sh (hunk 10 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh]: Oh I just tested and yeah it works.. changed it to that.. thanks :D

File: internal/core.go (hunk 7 above)
Comment (MOmarMiraj) [reply internal/core.go]: Now its done..

File: scripts/go_release.sh (hunk 8 above)
Comment (MOmarMiraj) [reply scripts/go_release.sh]: As we discussed, I added the two step release process.

I added a folder with all the changelogs for the release as it was the easier way to do it than creating one changelog file and appending the newest changelogs to the top of the file as well as cutting everything out besides the most recent release changelog.

File: Makefile [hunk 16]
```diff
@@ -0,0 +1,6 @@
+release:
+	scripts/release.sh
+
+prep-release:
+	scripts/prep-release.sh
+
```
Comment (hculea) [inline Makefile:6]: Let's make sure all the files have a proper new line, for POSIX compliance.

Comment (hculea) [reply Makefile:6]: There are also other files that do not, please have a look over those too.

File: internal/changelogs/README.md [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+
+Step 1. Make any changes to the SDK as required
+
+Step 2. Go to the root of the repo and run 
+```
+make prep-release
+```
+Follow the scripts instructions and the release has now been prepped.
+
+Step 3. Ensure that the correct files have been updated - i.e. version/build files, changelog has been added to this folder and has the correct file name and content
```
Comment (hculea) [inline internal/changelogs/README.md:11]: Suggest `git diff` to view the changes?

File: internal/changelogs/README.md [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+
+Step 4. Ensure your GITHUB_TOKEN environment variable is set as this will allow you to create the tags/release and push it.
+
+Step 5. If everything looks good, at the root of the repo, run this
+```
+make release
+```
```
Comment (hculea) [inline internal/changelogs/README.md:18]:
```suggestion
Step 5. If everything looks good, at the root of the repo, run:
```
make release
```
```

File: internal/changelogs/README.md [hunk 19]
```diff
@@ -0,0 +1,19 @@
+## How to Prepare a Release for the Go SDK
+
+Step 1. Make any changes to the SDK as required
```
Comment (hculea) [inline internal/changelogs/README.md:3]: Let's mention a few other things:
* script can be run from main and feature branch. If ran from main, a new branch will be created.
* one needs GitHub write permissions to the Go SDK repo to run this script

File: scripts/prep-release.sh [hunk 20]
```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
+
+# Helper script to prepare a release for the Go SDK.
+
+# Read the version number and build number from the respective files
+current_version_number=$(< internal/version)
+current_build_number=$(< internal/version-build)
```
Comment (hculea) [inline scripts/prep-release.sh:7]: these two shouldn't be required anymore, correct?

File: scripts/prep-release.sh [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+            echo "Please enter a build number in the 'Mmmppbb' format."
+        fi
+    done
+}
+
+# Update and Validate the version number
+validate_and_update_version_number
+
+# Update and Validate the build number
+update_and_validate_build_number
```
Comment (hculea) [inline scripts/prep-release.sh:56]: Let's name these both `update_and_validate` - or anything else consistent.

File: scripts/release.sh [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+# Add changes and commit/push to branch
+git add .
+git commit -m "Release for ${version}"
+git push origin ${branch}
+
+# Ensure GITHUB_TOKEN env var is set
+if [ -z "${GITHUB_TOKEN}" ]; then
+  echo "GITHUB_TOKEN environment variable is not set."
+  exit 1
+fi
```
Comment (hculea) [inline scripts/release.sh:40]: You'd want to validate this at the beginning of the script. We shouldn't tag without being able to complete the release.

File: scripts/prep-release.sh [hunk 23]
```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
+
+# Helper script to prepare a release for the Go SDK.
```
Comment (hculea) [inline scripts/prep-release.sh:3]: Let's make a check here that the working directory is clean (i.e. there are no uncommitted changes). This will save us someday from releasing some obscure WIP code

File: Makefile (hunk 16 above)
Comment (MOmarMiraj) [reply Makefile:6]: Done!

File: scripts/prep-release.sh (hunk 20 above)
Comment (MOmarMiraj) [reply scripts/prep-release.sh:7]: The current_version_number isn't required but the current_build_number is as later on in the script, we test to see if the build # has changed as we don't want the user to try to make a release with the same version.

File: internal/changelogs/README.md (hunk 19 above)
Comment (MOmarMiraj) [reply internal/changelogs/README.md]: Done!

File: scripts/prep-release.sh (hunk 23 above)
Comment (MOmarMiraj) [reply scripts/prep-release.sh:3]: Added!

File: scripts/prep-release.sh (hunk 21 above)
Comment (MOmarMiraj) [reply scripts/prep-release.sh]: Removed the number from both functions, but as one function updates and validates the version and the other does the build.

File: scripts/prep-release.sh [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+
+changelog_file="internal/changelogs/"${version}"-"${build}""
+
+# Store the changelog input into a file
+{
+   echo "Release Notes for: v"${version}""
+   echo ""
+   echo "${changelog_content}"
+   echo "[${build}]"
+   echo ""
```
[the remark this answers is not in the document]
Comment (MOmarMiraj) [reply scripts/prep-release.sh]: Done!

File: scripts/release.sh (hunk 22 above)
Comment (MOmarMiraj) [reply scripts/release.sh]: Good catch! Moved to the beginning of the script.

File: scripts/prep-release.sh (hunk 21 above)
Comment (hculea) [reply scripts/prep-release.sh]: My point is that one is called  "update and validate" and the other "validate and update".

**Comment by hculea** [comment]
Code looks good, but I would like a second look from @AndyTitu as well.

Additionally, can you please make the repository that you used for testing public, for us to also be able to functionally test?

Please update the description with the due testing steps.

File: scripts/prep-release.sh (hunk 21 above)
Comment (MOmarMiraj) [reply scripts/prep-release.sh]: Ohh.. I didn't realize lol fixed.

**Comment by hculea** [comment]
Approving the code, but I would love if you could functionally test together with @AndyTitu such that he can confirm that the flow is the same as the one he follows when releasing.

**Review by hculea — APPROVED** [review]

File: scripts/prep-release.sh [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+
+# Function to validate the build number format.
+# SEMVER Format: Mmmppbb - 7 Digits 
+update_and_validate_build() {
+    while true; do
+        # Prompt the user to input the build number
+        read -p "Enter the build number (format: Mmmppbb): " build
+
+        # Validate the build number format
+        if [[ "${build}" =~ ^[0-9]{7}$ ]]; then
```
Comment (AndyTitu) [inline scripts/prep-release.sh:46]: Nit: Feel free to skip for now, but we could also validate that the build number is not equal or less than the current build number.

File: internal/version
Comment (AndyTitu) [inline internal/version:1]: Wouldn't it be cleaner to have the following dir structure:

- internal
    -  release (new)
       - change-logs
       - README.md
       - version
       - version-build

File: scripts/prep-release.sh [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+# Read changelog input from the user until Ctrl+D is pressed
+changelog_content=""
+while IFS= read -r line; do
+    changelog_content+="${line}"$'\n' # Append each line to the variable with a newline character
+done
+
+changelog_file="internal/changelogs/"${version}"-"${build}""
+
+# Store the changelog input into a file
+echo "${changelog_content}" >> "${changelog_file}"
```
Comment (AndyTitu) [inline scripts/prep-release.sh:83]: Functionally tested and I got this output containing an error:

```
scripts/prep-release.sh: line 83: internal/changelogs/0.1.0-beta.12-0010012: No such file or directory
Release has been prepared..
Make sure to double check version/build numbers in their appropriate files and
changelog is correctly filled out.
Once confirmed, run 'make release' to release the SDK!
```
I know that the  `>>` operator should create a new file if it doesn't exist but in my case it doesn't

File: internal/core.go [hunk 27]
```diff
@@ -2,15 +2,18 @@ package internal
...
 )
 
 const (
-	SDKSemverVersion      = "0010009" // v0.1.0-beta.9
 	SDKLanguage           = "Go"
 	DefaultRequestLibrary = "net/http"
 )
 
+//go:embed version-build
+var SDKSemverVersion string
```
Comment (AndyTitu) [inline internal/core.go:15]: Bonus points for this smooth approach!

File: scripts/prep-release.sh [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+    echo "Build version hasn't changed. Stopping." >&2
+    exit 1
+fi
+
+echo "Enter your changelog for the release (press Ctrl+D when finished):"
+
+# Read changelog input from the user until Ctrl+D is pressed
+changelog_content=""
+while IFS= read -r line; do
+    changelog_content+="${line}"$'\n' # Append each line to the variable with a newline character
```
Comment (AndyTitu) [inline scripts/prep-release.sh:77]: Functional test edge case:

The instructions told me to press ctrl+d when I'm finished with my notes, so I did press it right after I wrote the text (I didn't add a new line to it)
```
➜  onepassword-sdk-go git:(main) make prep-release
scripts/prep-release.sh
Enter the version number (format: x.y.z(-beta.w)): 0.1.0-beta.11
New version number is: 0.1.0-beta.11
Enter the build number (format: Mmmppbb): 0010011
New build number is: 0010011
Enter your changelog for the release (press Ctrl+D when finished):
Release notes.^D
```
The ^D appeared at the end of my release notes. So I pressed ctrl+D again. Everything seemed fine but:

```
➜  onepassword-sdk-go git:(main) ✗ cat internal/changelogs/0.1.0-beta.11-0010011


```
The file was empty.

This works correctly though when I add a new line after my release notes.

File: scripts/release.sh [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+    git stash
+    git fetch origin
+    git checkout -b rc/"${version}"
+    git stash pop
+fi
+
+# Add changes and commit/push to branch
+git add .
+git commit -m "Release v${version}"
+git push origin ${branch}
```
Comment (AndyTitu) [inline scripts/release.sh:40]: You probably want to use this command here: `git push --set-upstream origin "${branch}"`

File: scripts/release.sh [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+if [[ "${branch}" = "main" ]]; then
+    git stash
+    git fetch origin
+    git checkout -b rc/"${version}"
+    git stash pop
+fi
+
+# Add changes and commit/push to branch
+git add .
+git commit -m "Release v${version}"
```
Comment (AndyTitu) [inline scripts/release.sh:39]: We should make sure the commit gets signed regardless of the environment which runs this script: 
```suggestion
git commit  -S -m "Release v${version}"
```

**Comment by AndyTitu** [comment]
Functional test unexpected behaviour:

```
➜  onepassword-sdk-go git:(main) ✗ make release
scripts/release.sh
Saved working directory and index state WIP on main: 8d41aeb Fix state
Switched to a new branch 'rc/0.1.0-beta.12'
On branch rc/0.1.0-beta.12
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   internal/version
	modified:   internal/version-build

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	internal/changelogs/0.1.0-beta.12-0010012

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (bd06505bc4b8fc597a1f0044b68030fb8e0b3898)
[rc/0.1.0-beta.12 04b3796] Release v0.1.0-beta.12
 3 files changed, 4 insertions(+), 2 deletions(-)
 create mode 100644 internal/changelogs/0.1.0-beta.12-0010012
Everything up-to-date
https://github.com/MOmarMiraj/onepassword-sdk-go/releases/tag/v0.1.0-beta.12
```

Some lines from this log look weird such as:

```
no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (bd06505bc4b8fc597a1f0044b68030fb8e0b3898)
```

The changes seem to have been committed from my inspections but my branch hasn't been pushed to the repo.

**Comment by AndyTitu** [comment]
Functionally tested both scripts thoroughly with all various input I could think of and the processes behave well apart from the 2 mentions. Also I left some code and dir structure suggestions

File: scripts/prep-release.sh (hunk 25 above)
Comment (MOmarMiraj) [reply scripts/prep-release.sh]: Done!

File: internal/version
Comment (MOmarMiraj) [reply internal/version:1]: Good Idea.. Added!

File: scripts/prep-release.sh (hunk 26 above)
Comment (MOmarMiraj) [reply scripts/prep-release.sh]: This should be inside the repo already so no problem here.

File: scripts/release.sh [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+# Get Current Branch Name
+branch="$(git rev-parse --abbrev-ref HEAD)"
+
+# if on main, then stash changes and create RC branch
+if [[ "${branch}" = "main" ]]; then
+    git stash
+    git fetch origin
+    git checkout -b rc/"${version}"
+    git stash pop
+fi
```
[the remark this answers is not in the document]
Comment (MOmarMiraj) [reply scripts/release.sh]: Great idea, I added this to the `prep-release.sh` script

File: scripts/release.sh (hunk 29 above)
Comment (MOmarMiraj) [reply scripts/release.sh]: Fixed

**Comment by MOmarMiraj** [comment]
> Functional test unexpected behaviour:
> 
> ```
> ➜  onepassword-sdk-go git:(main) ✗ make release
> scripts/release.sh
> Saved working directory and index state WIP on main: 8d41aeb Fix state
> Switched to a new branch 'rc/0.1.0-beta.12'
> On branch rc/0.1.0-beta.12
> Changes not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git restore <file>..." to discard changes in working directory)
> 	modified:   internal/version
> 	modified:   internal/version-build
> 
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
> 	internal/changelogs/0.1.0-beta.12-0010012
> 
> no changes added to commit (use "git add" and/or "git commit -a")
> Dropped refs/stash@{0} (bd06505bc4b8fc597a1f0044b68030fb8e0b3898)
> [rc/0.1.0-beta.12 04b3796] Release v0.1.0-beta.12
>  3 files changed, 4 insertions(+), 2 deletions(-)
>  create mode 100644 internal/changelogs/0.1.0-beta.12-0010012
> Everything up-to-date
> https://github.com/MOmarMiraj/onepassword-sdk-go/releases/tag/v0.1.0-beta.12
> ```
> 
> Some lines from this log look weird such as:
> 
> ```
> no changes added to commit (use "git add" and/or "git commit -a")
> Dropped refs/stash@{0} (bd06505bc4b8fc597a1f0044b68030fb8e0b3898)
> ```
> 
> The changes seem to have been committed from my inspections but my branch hasn't been pushed to the repo.

This is due to `git stash pop` as we didn't commit the changes that we popped, so this isn't really a error and is fine.

I just functionally tested the newest iteration and the branch is now being pushed.. don't know what the exact problem was but it works now and you can test in the test repo again.

File: scripts/prep-release.sh (hunk 28 above)
Comment (MOmarMiraj) [reply scripts/prep-release.sh]: I changed the method to adopt the same style as CLI. Should be good now!

**Comment by AndyTitu** [comment]
@MOmarMiraj My hunch is that the tests are failing due to the go:embed that now embeds the build version + new line, which makes for an invalid User Agent header

File: internal/release/scripts/release.sh [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+
+git tag -a -s  "v${version}" -m "${version}"
+
+# Get Current Branch Name
+branch="$(git rev-parse --abbrev-ref HEAD)"
+
+# Add changes and commit/push to branch
+git add .
+git commit -S -m "Release v${version}"
+git push --set-upstream origin ${branch}
```
Comment (AndyTitu) [inline internal/release/scripts/release.sh:32]: Also these were supposed to be moved along with the rc branch creation. When you prepare the release you'd want the release candidate to make it to remote before hitting release.

File: internal/release/scripts/prep-release.sh [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+
+# Get Current Branch Name
+branch="$(git rev-parse --abbrev-ref HEAD)"
+
+# if on main, then stash changes and create RC branch
+if [[ "${branch}" = "main" ]]; then
+    git stash
+    git fetch origin
+    git checkout -b rc/"${version}"
+    git stash pop
```
Comment (AndyTitu) [inline internal/release/scripts/prep-release.sh:86]:
```suggestion
    git stash apply
```

Should produce nicer log messages

**Review by AndyTitu — APPROVED** [review]
lgtm apart from my last 2 that should get addressed before merging

File: internal/release/scripts/release.sh (hunk 32 above)
Comment (MOmarMiraj) [reply internal/release/scripts/release.sh]: Fixed!
